apps/web/src/plugins/affine/index.tsx

```diff
@@ -140,9 +140,9 @@ export const AffinePlugin: WorkspacePlugin<WorkspaceFlavour.AFFINE> = {
       });
       await mutate(matcher => matcher === QueryKey.getWorkspaces);
     },
-    get: async workspaceId => {
+    get: async (workspaceId, opts) => {
```

Comment:
I think the correct implementation should be 

```ts
const workspaces = await loadFromIndexeddb()
// do not await remote fetch
loadFromCloud().then(remoteWorkspaces => {
  const result = merge(workspaces, remoteWorkspaces);
  // tell the upstream to rerender
  context.setWorkspaces(ws => merge(ws, result))
})

return workspaces 
```

Reply:
In this case, typeof Plugin should be replaced by `PluginFunction`

```ts
type PluginFunction = (context: AppContext) => Plugin
interface AppContext {
  setWorkspaces: Function
  // ...
}
```

Reply:
Some of my thoughts:

For data loading in two phases:
The current implementation does not think about load all workspaces all together. In the design, each workspace atom is a derived atom from its ID, where the ID is from different providers (merged); after we have ID (and its type), we will then load individual data by different plugin types and its value can be originated from multiple providers.

For introducing a context to trigger rerender:
I think the philosophy of WorkspacePlugin is just a upstream data fetcher, which implements how to load/mutate data without inferencing other states. Having a context here may make the data flow harder to be understood.

Ideally, I want to have a GET function which will emit value multiple times, so that we have the optimal data flow that reactively drives the UI. This is not possible with an async function, but is possible with an Observable (in Rxjs). Before  we are going to have a more reactive library, I think right now the possible solution is to have GET (+ List?) to return `Promise[]`, so that the caller of GET can be responsible for dealing with the multiple outputs.